common/channel/src/message_queues.rs

```diff
@@ -100,17 +107,25 @@ impl<K: Eq + Hash + Clone, T> PerKeyQueue<K, T> {
         if let Some(c) = self.counters.as_ref() {
             c.with_label_values(&["enqueued"]).inc();
         }
-        let max_queue_size = self.max_queue_size.get();
+
         let key_message_queue = self
             .per_key_queue
             .entry(key.clone())
-            .or_insert_with(|| VecDeque::with_capacity(max_queue_size));
+            // Only allocate a small initial queue for a new key. Previously, we
+            // allocated a queue with all `max_queue_size_per_key` entries;
+            // however, this breaks down when we have lots of transient peers.
+            // For example, many of our queues have a max capacity of 1024. To
+            // handle a single rpc from a transient peer, we would end up
+            // allocating ~ 96 b * 1024 ~ 64 Kib per queue.
+            .or_insert_with(|| VecDeque::with_capacity(1));
+
         // Add the key to our round-robin queue if it's not already there
         if key_message_queue.is_empty() {
             self.round_robin_queue.push_back(key);
         }
```

Comment:
I think there might be another race condition here if multi-threaded, where we could add a key to the round_robin_queue twice, which probably isn't a big deal.